### ISSUE TITLE
[prefer_adjacent_string_concatenation] should Not Warn for Non-string Left-hand type

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [dart-lang/linter#2188]
## Description

[prefer_adjacent_string_concatenation](https://dart-lang.github.io/linter/lints/prefer_adjacent_string_concatenation.html) does not consider the left-hand type when warning.  

The lint should prevent `string + string`, however, the lint appears for `any + string`.

## Reproduction

```dart

void main() {
  const a1 = '';

  // Correct prefer_adjacent_string_concatenation lint.
  print(a1 + '');

  const a2 = Custom();

  // Incorrect prefer_adjacent_string_concatenation lint.
  print(a2 + '');
}

class Custom {
  const Custom();

  Custom operator +(dynamic other) => const Custom();
}
```

https://dartpad.dev/461a20feb538a3c9d6fc71c1ab06f4d7

**Comment by asashour** [comment]
I am unable to reproduce with `master`.

Both cases are not triggering the lint.

The first one is very similar to

https://github.com/dart-lang/linter/blob/1e1300a9dcfaa3e4df1fe26fd00c3c5e44e81b77/test_data/rules/prefer_adjacent_string_concatenation.dart#L20

which should be `OK`

**Comment by creativecreatorormaybenot** [comment]
@asashour do you have the lint enabled? Otherwise, it might have been fixed / changed, I am not sure. Did you try exactly my sample with the lint rule applied?

**Comment by asashour** [comment]
I tested it in two ways

1. With copying the sample in `prefer_adjacent_string_concatenation.dart` in `master` and running `dart test -N prefer_adjacent_string_concatenation`

2. Running with a new project, copying the sample in `main.dart` and enabling the `prefer_adjacent_string_concatenation` lint, no lints found. When a `bad` case is added, the lint is triggered.

Version used:

     Dart SDK version: 2.14.0-321.0.dev (dev) (Thu Jul 15 08:55:11 2021 -0700) on "windows_x64"
